### DISCUSSION TITLE
Svelte port - technique

### DISCUSSION THREAD

**Discussion by ghost** [xyflow/xyflow#3012] [Svelte]
rewriting react-flow features for svelte framework is not scalable and heavy effort, for example, in future a better framework comes in, a typical response would be to rewrite the features that adapts to the framework.

I'm proposing a single source of truth and generating code targeting a particular framework. Idea is basically writing a template for component syntax targeting a particular framework.

I'll be glad to provide more details and for starter here it is 
https://github.com/imvetri/ui-editor/wiki open it and navigate to below section
![image](https://user-images.githubusercontent.com/6542274/233905400-3c6fb82f-d03f-4151-94f4-f57ab36f2d0f.png)

**Comment by moklick** [comment]
What is your point here? Are you suggesting that we should use ui-editor for create a framework agnostic version of React Flow?

**Reply by ghost** [reply]
I'm not proposing ui-editor, i brought it up to show and tell there is a technique to maintain reactflow for cross framworks consumers.

And I'd like to share techniques and philosophy I learnt from there so that this project can have a framework agnostic solution.

In a way, react-flow will have wider consumers, with less maintained effort from your team.

**Reply by moklick** [reply]
Got it, thanks. How is it better than / different from https://github.com/BuilderIO/mitosis ?

**Reply by ghost** [reply]
Looks interesting!. Thanks for sharing it.

**Reply by ghost** [reply]
Started a thread here https://github.com/BuilderIO/mitosis/discussions/1152

**Reply by ghost** [reply]
@moklick I got carried away looking at their setup. its cool, however, keeping or defining the user interface behaviour, even with mitosis, we are required to write the components definition and behaviour in their syntax.

What I was trying to propose is a design interaction language, or a standard that will define how a component behaves. For example, designer's term of `variant` is used to create components. My proposal is to bring in a component composition rules. I'll be glad to provide more example, if you would like to know.

I noticed this specifically when I had a look at how handles are composed part of a node.

cc @Elringus

**Reply by elringus** [reply]
I don't think locking to yet another framework/language/whatever is a good idea. Just keep the core stuff in vanilla js (preferably w/o 3rd party deps) and maintain wrappers for 1 or 2 most popular frameworks. This way you're proofed against flaky framework trends and open for community to maintain wrappers for less popular frameworks.

**Comment by elringus** [comment]
+1 to the idea of making a vanilla JS core with a react/svelte/whatever wrapper instead of hard-wiring into specific framework. Frameworks come and go, you never know what would replace React in the future: it may be Svelte or it may be SolidJS or something else. It's easier to replace a wrapper than rewrite the whole thing over.

**Reply by moklick** [reply]
Our solution is somewhere in between. We have a vanilla core with functions for:

* zoom and pan
* dragging nodes and selections
* handle functionality 
* interactive minimap 
* edge paths creation 
* some utils
* types

These things can be used by the libraries but we are still able to build Svelte/Solid/Vue libraries that can use the advantages of the specific framework.

**Reply by ghost** [reply]
I had this strange idea this was the thought 

react-flow helps to create **flows from code**. why not create **code from flow**. 

In other words, to **emphasis** more on the **concept of flow,** defining theories, philosophy, defining fundamental rules. I know that we are defining the concept of flow in terms of node, edges, which for some reason does not make much catch sense. I believe there is more to the concept of flow as it is a very highly abstracted and can be applied in many computation, but involving code in it, distracts from the concept itself.

Background - I noticed just like react-flow there is another project mimicking yours in svelete-flow. I'm not sure if it is yours or someone else, but as a general thought it looks like copying a concept in a different framework which is not efficient use of our time. 

Just sharing this because I thought, its time for project to innovate, get creative because technical aspects of implementation can be achieved some how, but knowing what we are building needs to degenerate into more concrete form. I used best choice of words to explain what I was thinking, feel free to question or ask more if I didnt communicate it clear
